### PR TITLE
SIP-16: Add more tx_context

### PR DESCRIPTION
**Pull request by EasonC13**

### CODE REVIEW

File: sips/sip-temporary_title.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+Regarding the use of tx_context::is_signed_by(ctx, address), we plan to run some transaction that need multiple signers to proceed.
+
+## Specification
+
+`tx_context::gas_sender(ctx)` return `address` of the gas sender.
+
+`tx_context::is_signed_by(ctx, address)` return `boolean`.
+
+`tx_context::signers(ctx)` return `vector<address>`.
```
Comment (sblackshear) [inline sips/sip-temporary_title.md:37]: Couple of edge cases to think about here:
- What order does it return the addresses in? I would vote for gas sender last, but it doesn't matter as long as it's clear/consistent
- This should always consist of distinct addresses, right

**Comment by longbowlu** [comment]
> Therefore, we aim to delete objects that have been sponsored to create by us in order to collect storage rebates and prevent individuals from abusing sponsored transactions to create objects with sponsor and then deleting them using their own addresses to get rebate from sponsor.

Could you elaborate on how these new context would be used to achieve this motivation? an example function will be very helpful. thanks

File: sips/sip-temporary_title.md (hunk 1 above)
Comment (EasonC13) [reply sips/sip-temporary_title.md:37]: > What order does it return the addresses in?

At SDK, When we run `executeTransactionBlock` we pass the signature in as an array. Maybe the order can be the same as the array.

```
result = await suiClient.executeTransactionBlock({
        transactionBlock: sponsorSignedTxBlockBytes,
        signature: [sponsorSignature, clientSignature],
        options: {
          showEffects: true,
          showEvents: true,
          showObjectChanges: true,
        },
      });
```

**Comment by admin-aftermath** [comment]
The ability to retrieve the signer (or signers) is something we have been looking for as well!

**Comment by EasonC13** [comment]
> The ability to retrieve the signer (or signers) is something we have been looking for as well!
> 
> > Could you elaborate on how these new context would be used to achieve this motivation?
> 
> I assume a flow like this would get their job done:
> 
> ```
> // During creation
> if (is_sponsored(...)) {
>     set_creator(OBJECT, option::some(SPONSOR))
> } else {
>     set_creator(OBJECT, option::none())
> }
> 
> // During deletion
> if (option::is_some(creator(OBJECT))) {
>     assert(option::extract(creator(OBJECT)) == tx_context::sender(ctx), EInvalidSender
> }
> ```

Thank you, that is a good example

**Comment by EasonC13** [comment]
> I would probably name the "gas_sender" function "gas_owner" instead, but otherwise, the new list of APIs sounds reasonable to me.

Agreed, `gas_owner` is better.

**Comment by nathanramli** [comment]
Is there a scenario where multiple signers, aside from the `sender` and `gas_owner`, might be involved?

**Comment by EasonC13** [comment]
> Is there a scenario where multiple signers, aside from the `sender` and `gas_owner`, might be involved?

Yes if we use multisig feature:

https://docs.sui.io/concepts/cryptography/transaction-auth/multisig

**Comment by lxfind** [comment]
> Yes if we use multisig feature

Note that even with multisig, there is still one sender address which is the combined address.
Today we only support at most two signers in a transaction, the sender and gas owner.

**Comment by SA124** [comment]
@EasonC13 

Thanks for your SIP proposal. In order to process the SIP, the SIP Editor requires write access to the forked repository in order to change the status and add additional information.

Could you please grant write access on the forked repository to the following GitHub accounts?

SA124 
hyd628

Once this has been completed, we can move the proposal forward.

Best regards, 
Shereen

**Comment by EasonC13** [comment]
> @EasonC13
> 
> Thanks for your SIP proposal. In order to process the SIP, the SIP Editor requires write access to the forked repository in order to change the status and add additional information.
> 
> Could you please grant write access on the forked repository to the following GitHub accounts?
> 
> SA124 hyd628
> 
> Once this has been completed, we can move the proposal forward.
> 
> Best regards, Shereen

Invite sent.

**Comment by nathanramli** [comment]
> > Yes if we use multisig feature
> 
> Note that even with multisig, there is still one sender address which is the combined address. Today we only support at most two signers in a transaction, the sender and gas owner.

Yes, this is what I meant. Even if we can obtain the list of signers, the sender and the signers could be different. The sender isn't necessarily a signer, although currently, we treat the sender as a signer.

**Comment by tnowacki** [comment]
Just as a quick update, we plan on adding `gas_sender`, but are still debating some of the details for something like `is_signed_by` or `signers`

**Review by hyd628 — COMMENTED** [review]
Looks good. Moving to Review status.

**Review by hyd628 — DISMISSED** [review]

**Comment by EasonC13** [comment]
Hi, want to know the follow up about this.

**Comment by wriches** [comment]
Status changed to Stagnant due to inactivity. The SIP can be revived at a later date.

**Comment by Euraxluo** [comment]
I would like to ask if you can add a signature, such as zklogin for a special identity

**Review by wriches — APPROVED** [review]

**Comment by wriches** [comment]
SIP withdrawn due to delay. A new SIP can be resubmitted in the future.

**Comment by Sceat** [comment]
What would be needed to land this SIP ? @wriches what do you mean by delay ?

**Comment by Sceat** [comment]
@EasonC13 maybe you could re-submit the SIP ?

**Pull request merged**